velox/exec/Task.cpp

```diff
@@ -563,178 +563,178 @@ void Task::start(
+  // first.
+  if (numDriversUngrouped_ > 0) {
+    // Create the drivers we are going to run for this task.
+    std::vector<std::shared_ptr<Driver>> drivers;
+    drivers.reserve(numDriversUngrouped_);
+    createSplitGroupStateLocked(kUngroupedGroupId);
+    createDriversLocked(kUngroupedGroupId, drivers);
+    if (pool_->stats().currentBytes != 0) {
+      VELOX_FAIL(
+          "Unexpected memory pool allocations during task[{}] driver initialization: {}",
+          taskId_,
+          pool_->treeMemoryUsage());
     }
 
-    std::unique_lock<std::mutex> l(self->mutex_);
-
-    // Preallocate a bunch of slots for max concurrent grouped execution
-    // drivers, if needed.
-    if (self->numDriversPerSplitGroup_ > 0) {
-      self->drivers_.resize(
-          self->numDriversPerSplitGroup_ * self->concurrentSplitGroups_);
+    // Prevent the connecting structures from being cleaned up before all
+    // split groups are finished during the grouped execution mode.
+    if (isGroupedExecution()) {
+      splitGroupStates_[kUngroupedGroupId].mixedExecutionMode = true;
     }
 
-    // We create the drivers running pipelines in ungrouped execution mode
-    // first.
-    if (self->numDriversUngrouped_ > 0) {
-      // Create the drivers we are going to run for this task.
-      std::vector<std::shared_ptr<Driver>> drivers;
-      drivers.reserve(self->numDriversUngrouped_);
-      self->createSplitGroupStateLocked(kUngroupedGroupId);
-      self->createDriversLocked(self, kUngroupedGroupId, drivers);
-      if (self->pool_->stats().currentBytes != 0) {
-        VELOX_FAIL(
-            "Unexpected memory pool allocations during task[{}] driver initialization: {}",
-            self->taskId_,
-            self->pool_->treeMemoryUsage());
+    // We might have first slots taken for grouped execution drivers, so need
+    // to append the ungrouped execution drivers afterwards in that case.
+    if (drivers_.empty()) {
+      drivers_ = std::move(drivers);
+    } else {
+      drivers_.reserve(drivers_.size() + numDriversUngrouped_);
+      for (auto& driver : drivers) {
+        drivers_.emplace_back(std::move(driver));
       }
+    }
 
-      // Prevent the connecting structures from being cleaned up before all
-      // split groups are finished during the grouped execution mode.
-      if (self->isGroupedExecution()) {
-        self->splitGroupStates_[kUngroupedGroupId].mixedExecutionMode = true;
+    // Set and start all Drivers together inside 'mutex_' so that
+    // cancellations and pauses have well defined timing. For example, do not
+    // pause and restart a task while it is still adding Drivers. If the given
+    // executor is folly::InlineLikeExecutor (or it's child), since the
+    // drivers will be executed synchronously on the same thread as the
+    // current task, so we need release the lock to avoid the deadlock.
+    if (dynamic_cast<const folly::InlineLikeExecutor*>(
```

Comment:
It would be great to remove this usage and assert this is no the case. Task::next is the API to use for single-threaded execution.